sklearn/linear_model/_glm/glm.py

```diff
@@ -20,13 +24,746 @@
+            sample_weight=sample_weight,
+            l2_reg_strength=self.l2_reg_strength,
+            n_threads=self.n_threads,
+            gradient_out=self.gradient,
+            hessian_out=self.hessian,
+            raw_prediction=self.raw_prediction,  # this was updated in line_search
+        )
+
+    def line_search(self, X, y, sample_weight):
+        # Use R' instead of X
+        super().line_search(X=self.R.T, y=y, sample_weight=sample_weight)
+
+    def check_convergence(self, X, y, sample_weight):
+        # Use R' instead of X
+        super().check_convergence(X=self.R.T, y=y, sample_weight=sample_weight)
+
+    def finalize(self, X, y, sample_weight):
+        n_features = X.shape[1]
+        w, intercept = self.linear_loss.weight_intercept(self.coef)
+        self.coef_original[:n_features] = self.Q @ w
+        if self.linear_loss.fit_intercept:
+            self.coef_original[-1] = intercept
+        self.coef = self.coef_original
+
+
+class LSMRNewtonSolver(NewtonSolver):
+    """LSMR based Newton solver.
+
+    The inner solver uses LSMR after the Newton update is cast into the IRLS
+    formulation. This means
+
+        H @ coef_newton = -G
+
+    with
+
+        G = X.T @ g + l2_reg_strength * coef
+        H = X.T @ diag(h) @ X + l2_reg_strength * identity
+        g = loss.gradient = pointwise gradient
+        h = loss.hessian = pointwise hessian
+
+    is cast as a least squares problem
+
+        min ||A @ newton_step - b||_2^2
+
+    with
+
+        A = [                    sqrt(h) * X]
+            [sqrt(l2_reg_strength) * sqrt(P)]
+        b = [                                - g / sqrt(h)]
+            [- sqrt(l2_reg_strength) * sqrt(P) @ self.coef]
+        P = penalty matrix in 1/2 w @ P @ w,
+            for a pure L2 penalty without intercept it equals the identity matrix.
+
+    Note that this solver can naturally deal with sparse X.
+    """
+
+    def setup(self, X, y, sample_weight):
+        """Setup.
+
+        If None, initializes:
+            - self.coef
+        Sets:
+            - self.raw_prediction
+            - self.loss_value
+            - self.g = pointwise gradient
+            - self.h = pointwise hessian
+            - self.gradient
+            - self.sqrt_P = sqrt(l2_reg_strength) * sqrt(P)
+            - self.A_norm
+            - self.X = X
+        """
+        super().setup(X=X, y=y, sample_weight=sample_weight)
+        (self.g, self.h,) = self.linear_loss.base_loss.init_gradient_and_hessian(
+            n_samples=X.shape[0], dtype=X.dtype
+        )
+        # For a general (symmetric) penalty matrix P, we can use any square root of it,
+        # e.g. the Cholesky decomposition. Here, we can use the identity matrix and
+        # handle the intercept.
+        n_features = X.shape[1]
+        # We use a 1-d array instead of a 2-d diagonal array to save memory. We also
+        # simply omit to add a zero at the end for the intercept.
+        self.sqrt_P = np.full(
```

Comment:
Maybe this would be renamed to `self.sqrt_P_diag` to be explicit that the `(n_features, n_features)` `P` matrix is not needed.

Reply:
I agree that this would be a better name. But it would cause several new line breaks and make code less readable. This is already the case with `self.l2_reg_strength` which is way to long for what it's worth, IMHO.

Reply:
Personnally I prefer the extra line breaks :)

But this is minor so do as you wish and feel free to mark this thread resolved.